table/block.cc

```diff
@@ -505,9 +532,81 @@ bool IndexBlockIter::PrefixSeek(const Slice& target, uint32_t* index) {
+// the iterator will simply be set as "invalid", rather than returning
+// the key that is just pass the target key.
+void DataBlockIter::HashSeek(const Slice& target) {
+  assert(data_block_hash_index_);
+  Slice user_key = ExtractUserKey(target);
+  uint8_t entry = data_block_hash_index_->Seek(user_key);
+
+  if (entry == kNoEntry) {
+    current_ = restarts_;  // not found, Invalidate the iterator
+    return;
+  }
+
+  if (entry == kCollision) { // HashSeek not effective
+    status_ = Status::NotSupported("Hash Collision");
+    return;
+  }
+
+  uint32_t restart_index = entry;
+
+  // check if the key is in the restart_interval
+  SeekToRestartPoint(restart_index);
+  while (true) {
+    // Here we only linear seek the target key inside the restart interval.
+    // If a key does not exist inside a restart interval, we avoid
+    // further searching the block content accross restart interval boundary.
+    //
+    // TODO(fwu): check the left and write boundary of the restart interval
+    // to avoid linear seek a target key that is out of range.
+    //
+    // If using hash, we only linear search within the restart inteval.
+    if (!ParseNextDataKey(true /*within_restart_interval*/) ||
+        Compare(key_, target) >= 0) {
+      // we stop at the first potential matching user key.
+      break;
+    }
+  }
+
+  if ((current_ != restarts_) /* valid */ &&
+      // If the user key portion match do we consider key_ matches
+      user_comparator_->Compare(key_.GetUserKey(), user_key) == 0) {
+
+    // Here we are conservative and only support a limited set of cases
+    ValueType value_type = ExtractValueType(key_.GetKey());
+    if (value_type != ValueType::kTypeValue &&
+        value_type != ValueType::kTypeDeletion) {
+      status_ = Status::NotSupported("record type not supported");
+      return; // found, but not supported.
+    }
+
+    // Currently we do not fully support searching a key at specify snapshot.
+    // HashSeek only examine the first matched user_key. If the seqno is
+    // higher than the targe snapshot seqno, we just fall back to BinarySeek,
+    // without search further for the correct seqno.
+    uint64_t target_seqno = GetInternalKeySeqno(target);
+    uint64_t seqno = GetInternalKeySeqno(key_.GetKey());
+    if (target_seqno < seqno) {
+      status_ = Status::NotSupported("snapshot not fully supported");
+      // found, but not supported
+    }
+    return; // found
+  }
+  current_ = restarts_;  // not found, Invalidate the iterator
+}
+
 uint32_t Block::NumRestarts() const {
   assert(size_ >= 2*sizeof(uint32_t));
-  return DecodeFixed32(data_ + size_ - sizeof(uint32_t));
+  uint32_t block_footer = DecodeFixed32(data_ + size_ - sizeof(uint32_t));
+  return block_footer & 0x7FFFFFFF;
+}
+
+uint32_t Block::IndexType() const {
+  assert(size_ >= 2 * sizeof(uint32_t));
+  uint32_t block_footer = DecodeFixed32(data_ + size_ - sizeof(uint32_t));
+  return block_footer >> 31;
```

Comment:
Try to think about a way to consolidate this ">>31" and previous "block_footer & 0x7FFFFFFF". Ideally we process them with the same const.